### DISCUSSION TITLE
Servo Motor , PWM can't set  50Hz

### DISCUSSION THREAD

**Discussion by sangyuxiaowu** [nanoframework/Home#1230] [Q&A]
Hi,
I'm trying to use my board to control the servos, according to this article https://docs.nanoframework.net/devicesdetails/ServoMotor/README.html

```
Configuration.SetPinFunction(2, DeviceFunction.PWM16);
PwmChannel pwmChannel = PwmChannel.CreateFromPin(2, 50);
ServoMotor servoMotor = new ServoMotor(pwmChannel);
servoMotor.Start();

servoMotor.WritePulseWidth(0);
Thread.Sleep(2000);
servoMotor.WritePulseWidth(90);
Thread.Sleep(2000);
servoMotor.WritePulseWidth(180);
Thread.Sleep(Timeout.Infinite);
```

I'm sure there is no problem with the wiring, but this error is reported when creating the PwmChannel. It seems that it cannot be set to 50. After testing, it is found that it needs to be set to at least 2432

```text
引发的异常:“System.ArgumentException”(位于 System.Device.Pwm.dll 中)
    ++++ Exception System.ArgumentException - 0xfd000000 (1) ++++
    ++++ Message: 
    ++++ System.Device.Pwm.PwmChannel::NativeSetDesiredFrequency [IP: 0000] ++++
    ++++ System.Device.Pwm.PwmChannel::set_Frequency [IP: 0005] ++++
    ++++ System.Device.Pwm.PwmChannel::CreateFromPin [IP: 0026] ++++
    ++++ MotionControl.Program::Main [IP: 0039] ++++
```

board brushed FEATHER_S2 firmware. servos is SG92R
information:
https://www.waveshare.com/wiki/ESP32-S2-Pico

It's a strange question, because I'm just a hobbyist, I may not know some basic knowledge.

Any help at all appreciated!

thanks

**Comment by sangyuxiaowu** [comment]
I tested MicroPython again, which can be controlled normally.

**Reply by Ellerbach** [reply]
Are you using the channel? It is different than the GPIO number. Setting the GPIO pin needs to be done in coordination with the channel number. Check the Feather documentation.
On Feather the default is already the same number. So skip the pin setting in your code.
See https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/api-reference/peripherals/ledc.html

**Reply by sangyuxiaowu** [reply]
Yes, I am using the above code. But if `Configuration.SetPinFunction` is skipped then `PwmChannel` will return null.
I'm very sorry, maybe I didn't understand you, I'm a newbie.

**Reply by Ellerbach** [reply]
> I'm very sorry, maybe I didn't understand you, I'm a newbie.

Don't be sorry :-)

So try `Configuration.SetPinFunction(2, DeviceFunction.PWM2);` as for Feather, the Pin number = the Pwm Channel.
We'll find out :-) And adjust the docuementation. Feather S2 has been added recently.

**Reply by sangyuxiaowu** [reply]
I tried Configuration.SetPinFunction(2, DeviceFunction.PWM2); before it became 16 because I tried every one from 1-16. Also tried PwmChannel.Create and PwmChannel.CreateFromPin. Seems to be reporting the same error. I also tried to ask ChatGPT and New Bing, but they didn't solve it. I also asked my friends around me, and simulated PWM, but it seems to be useless. I am a little desperate for this problem.
